### ISSUE TITLE
Outputting larger images and getting around memory constraints

### ISSUE THREAD

**Issue by ikarth** [alexjc/neural-doodle#75]
The biggest barrier I see to adopting this for practical use right now is the size of the image output. I can find lots of applications today for 1080p output, but it's harder for 400x400. I can run on the CPU, at much slower speeds, but being able to run on the GPU at a quarter of the current speed would be sufficiently fast to get started.

I don't know enough yet about what's going on under the hood to know how feasible this is with the current codebase. 

The barriers I see right now:

- Breaking the larger image into tiles causes artifacting along the edges of the tiles

- Smaller tiles also seem to produce a different result; the scale affects the outcome and the results aren't consistent

- There are also differences across tiles; are the images being normalized before processing?

- Using smaller style images on larger content images seems to crash the GPU. (See Issue #74)

![woods_poster3](https://cloud.githubusercontent.com/assets/1009991/14401380/4bee47a6-fdde-11e5-8e65-15cbcb241703.png)
Tiled image, GPU, six 256x256 tiles

![woods_u_x](https://cloud.githubusercontent.com/assets/1009991/14401379/424866a0-fdde-11e5-93b1-97d3abf0aee5.png)
Non-tiled Image, CPU

Now, if the entire neural net needs to be loaded into memory to process even part of the image then dividing the image into sections likely doesn't help. If, on the other hand, the image can be meaningfully processed in sections, then doodle.py can divide the processing into sections. However, splitting the images before running them through doodle.py doesn't currently get consistant enough results to split the task that way.

**Comment by alexjc** [comment]
I agree entirely with your assessment, I've been working on something already.

For your solution, you should try overlapping the tiles and blending them together to hide the edges! This would be better already. There's no contrast adjustment.  For smaller tiles you can adjust the number of resolutions/phases processed too (e.g. make it one less if you divide into two tiles).

**Comment by ikarth** [comment]
These are from 4/10, so they use the older codebase, but they might help other people thinking about doing tiling:

[tile_test_five_even](https://cloud.githubusercontent.com/assets/1009991/14762299/131c0dc4-0945-11e6-87e4-b8f2704fd444.png)
[tile_test_five_odd](https://cloud.githubusercontent.com/assets/1009991/14762344/3b01f8b6-0946-11e6-9059-b635972f46d7.png)
![tile_test_five](https://cloud.githubusercontent.com/assets/1009991/14762349/498189a6-0946-11e6-9236-9cb278ede2df.jpg)

I tried stripped tiles first, since there was a naive way to combine alternating overlapping rows, at the cost of basically rendering the entire image twice. This level of overlap is likely unnecessary in most cases: the most obvious discontinuities tend to be in a halo five or six pixels wide around the edges.
The gap you see is where the strips ran out of GPU memory. (I've since also managed to run out of RAM when rendering on the CPU, so there are limits even there.)

Using strips that have a different aspect ratio than the original image seems to have resulted in a very different output. The other factor I've noticed when working with large images is that you may need a more detailed semantic map, since the scale of the features it finds appears to be affected (compare with the non-tiled all-on-CPU image above). This may partially be because an individual tile may have only one or two of the semantic regions, rather than the full set.

![compose_even_test3](https://cloud.githubusercontent.com/assets/1009991/14762419/b9631dec-0947-11e6-956e-e2ba591be604.png)
This one is a non-overlapping subset of tiles extracted from a larger set of overlapping tiles. These tiles preserve the aspect ratio of the source photo, but still exhibit the feature scaling issue. 

Also note that one of the tiles is noticeably darker: the semantic map for that one is almost entirely 'sky', with only a bit of one other color intruding, than seems to have resulted in a different intensity. You can notice some similar intensity differences, though more subtly, on the stripped tiles above.

These were all done with a separate python script that called the doodle script after chopping up the image with ImageMagick, so it was a very naive approach that ignored the commonalities between the tiles. 

One thing I have noticed, separate from the tiling experiments, is that the process is very good at populating edges and transitions with relevant content. Rather than just having a tree and a sky, consider also having an edge-of-tree-with-sky semantic annotation; the algorithm is unexpectedly good at giving transitions a convincing appearance. It has a little more trouble picking up shadow or direction on, say, a river bank, so a left-diagonal-edge and a right-diagonal-edge annotation may give you better results there.

I'm eager to try out the patch variety code, which I suspect will address some of the uniform feature scaling larger images exhibit.

**Comment by alexjc** [comment]
Very cool, thanks for the report!

The patch variety may affect your tiling in a negative manner—edges may not match up as well. I'm still working on the code for higher resolutions, may take a few more hours/days ;-)

**Comment by ikarth** [comment]
At the moment I'm more concerned with looking more interesting rather than edge-matching, since I don't need perfectly tiling segments right away, I just want them! But I'll have to try it and see. My more immediate project is to experiment with a wide range of small images and test out where the limitations and hidden opportunities are that I haven't discovered yet.
